aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -354,27 +358,123 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
 // add(Tensor, SparseTensor, Scalar)
 //    formerly known as spcadd
 // --------------------------------------------------------------------
-
 template <typename scalar_t>
-void add_dense_sparse_worker_cpu(Tensor& r, Scalar value, const SparseTensor& sparse, const Tensor& indices, const Tensor& values) {
+void add_dense_sparse_worker_scalar_cpu(Tensor& r, Scalar value, const SparseTensor& sparse, const Tensor& indices, const Tensor& values) {
   auto indices_accessor = indices.accessor<int64_t, 2>();
   auto values_accessor = values.accessor<scalar_t, 1>();
 
   scalar_t* r_ptr = r.data_ptr<scalar_t>();
   scalar_t cast_value = value.to<scalar_t>();
-
+  auto sparse_dim = sparse.sparse_dim();
+  std::vector<int64_t> result_stride(sparse_dim);
+  for (int64_t d = 0; d < sparse_dim; d++) {
+    result_stride[d] = r.stride(d);
+  }
   at::parallel_for(0, sparse._nnz(), 0, [&](int64_t start, int64_t end) {
     for (auto k = start; k < end; k++) {
       int64_t index = r.storage_offset();
-      for (int64_t d = 0; d < sparse.sparse_dim(); d++) {
-        index += r.stride(d) * indices_accessor[d][k];
+      for (int64_t d = 0; d < sparse_dim; d++) {
+        index += result_stride[d] * indices_accessor[d][k];
       }
       r_ptr[index] += cast_value * values_accessor[k];
     }
   });
 }
 
-Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTensor& sparse_, Scalar value) {
+template <typename scalar_t>
+inline void add_dense_sparse_worker_cpu(Tensor& r, Scalar value, const SparseTensor& sparse, const Tensor& indices, const Tensor& values) {
+
+  // Get the dense dimension element numbers of hybrid sparse tensor
+  int64_t values_dense_size = values.stride(0);
+  AT_ASSERT(values.is_contiguous());
+  scalar_t* v_ptr = values.data_ptr<scalar_t>();
+  v_ptr += values.storage_offset();
+
+  scalar_t* r_ptr = r.data_ptr<scalar_t>();
+  AT_ASSERT(r_ptr != nullptr);
+  r_ptr += r.storage_offset();
+
+  auto indices_accessor = indices.accessor<int64_t, 2>();
+  scalar_t cast_value = value.to<scalar_t>();
+  auto sparse_dim = sparse.sparse_dim();
+  std::vector<int64_t> result_stride(sparse_dim);
+  for (int64_t d = 0; d < sparse_dim; d++) {
+    result_stride[d] = r.stride(d);
+  }
+
+  at::parallel_for(0, sparse._nnz(), 0, [&](int64_t start, int64_t end) {
+    for (auto k = start; k < end; k++) {
+      auto r_index = r_ptr;
+      for (int64_t d = 0; d < sparse_dim; d++) {
+        r_index += result_stride[d] * indices_accessor[d][k];
+      }
+      auto v_index = v_ptr + k * values_dense_size;
+      PRAGMA_SIMD
+      for (auto v=0; v < values_dense_size; v ++) {
+        r_index[v] += cast_value * v_index[v];
+      }
+    }
+  });
+}
+
+static constexpr unsigned add_dense_sparse_max_lock_num_bits = 14;
+template <typename scalar_t>
+inline void add_dense_sparse_worker_critical_section_cpu(Tensor& r, Scalar value,
```

Comment:
Do you observe real models using this path much? It is much more easier (code-wise) to coalesce sparse tensor. And much more efficient performance wise to coalesce it once upon construction.

Reply:
yes, all [DLRM](https://github.com/facebookresearch/dlrm) add_out_dense_sparse_cpu will enter this branch now. Currently sparse_.coalesce() can't do in parallel which make it's spend more time than using the critical section to add out  without do coalescing first